From 9ceea22d9f7594c34412c1ea0caab8e38fa5b18e Mon Sep 17 00:00:00 2001
From: Nobuyoshi Nakada <nobu@ruby-lang.org>
Date: Thu, 24 Mar 2022 01:42:06 +0900
Subject: [PATCH] [Bug #18651] Pass undefined region in CP51932 as is to
 CP50220

The CP50220 encoder held a CP51932 byte pair in the undefined region
(0x8E followed by 0xE0..0xFE) as a katakana that might take a sound
mark, and later flushed it through tbl0208 at (c - 0x21) * 2 with c up
to 0x7E, reading up to 62 bytes past the end of the 126-byte table into
the output.

Backport to 3.3 of the series that fixed this on master, all of which
is in 3.4:

  a8d600edff Extract `iso2022jp_put_state`
  546f13fded Define `tbl0208` as a static array
  75a8c937e7 [Bug #18651] Pass undefined region in CP51932 as is to CP50220
  aa2662d590 Simplify offset calculations of `tbl0208`

75a8c937e7 is the fix; the other three are refactors its context and
the later CP50220 max_output fix depend on. 3.3 already marks tbl0208
with RBIMPL_ATTR_NONSTRING() for GCC 15 (b42afa1dbc, Backport #21497).
The conflicts in 546f13fded and aa2662d590 keep that, using
RBIMPL_ATTR_NONSTRING_ARRAY() for the final two-dimensional table as 3.4
does (846bb76075, Backport #21629). The resulting enc/trans/iso2022.trans
is identical to v3_4_11's.
---
 enc/trans/iso2022.trans     | 147 ++++++++++++++++++------------------
 test/ruby/test_transcode.rb |   2 +
 2 files changed, 74 insertions(+), 75 deletions(-)

diff --git a/enc/trans/iso2022.trans b/enc/trans/iso2022.trans
index ea7d2c8aec..bc42bbc19c 100644
--- a/enc/trans/iso2022.trans
+++ b/enc/trans/iso2022.trans
@@ -80,6 +80,34 @@ iso2022jp_init(void *statep)
     return 0;
 }
 
+static unsigned char *
+iso2022jp_put_state(unsigned char *sp, unsigned char *o, int oldstate, int newstate)
+{
+    if (oldstate != newstate) {
+        *o++ = 0x1b;
+        switch (newstate) {
+          case G0_ASCII:
+            *o++ = '(';
+            *o++ = 'B';
+            break;
+          case G0_JISX0201_KATAKANA:
+            *o++ = '(';
+            *o++ = 'I';
+            break;
+          case G0_JISX0208_1978:
+            *o++ = '$';
+            *o++ = '@';
+            break;
+          default:
+            *o++ = '$';
+            *o++ = 'B';
+            break;
+        }
+        *sp = newstate;
+    }
+    return o;
+}
+
 static VALUE
 fun_si_iso2022jp_decoder(void *statep, const unsigned char *s, size_t l)
 {
@@ -155,24 +183,7 @@ fun_so_iso2022jp_encoder(void *statep, const unsigned char *s, size_t l, unsigne
     else
         newstate = G0_JISX0208_1983;
 
-    if (*sp != newstate) {
-        if (newstate == G0_ASCII) {
-            *o++ = 0x1b;
-            *o++ = '(';
-            *o++ = 'B';
-        }
-        else if (newstate == G0_JISX0208_1978) {
-            *o++ = 0x1b;
-            *o++ = '$';
-            *o++ = '@';
-        }
-        else {
-            *o++ = 0x1b;
-            *o++ = '$';
-            *o++ = 'B';
-        }
-        *sp = newstate;
-    }
+    o = iso2022jp_put_state(sp, o, *sp, newstate);
 
     if (l == 1) {
         *o++ = s[0] & 0x7f;
@@ -203,10 +214,7 @@ finish_iso2022jp_encoder(void *statep, unsigned char *o, size_t osize)
     if (*sp == G0_ASCII)
         return 0;
 
-    *o++ = 0x1b;
-    *o++ = '(';
-    *o++ = 'B';
-    *sp = G0_ASCII;
+    o = iso2022jp_put_state(sp, o, *sp, G0_ASCII);
 
     return o - output0;
 }
@@ -400,24 +408,7 @@ fun_so_cp5022x_encoder(void *statep, const unsigned char *s, size_t l,
     else
         newstate = G0_JISX0208_1983;
 
-    if (*sp != newstate) {
-        if (newstate == G0_ASCII) {
-            *o++ = 0x1b;
-            *o++ = '(';
-            *o++ = 'B';
-        }
-        else if (newstate == G0_JISX0201_KATAKANA) {
-            *o++ = 0x1b;
-            *o++ = '(';
-            *o++ = 'I';
-        }
-        else {
-            *o++ = 0x1b;
-            *o++ = '$';
-            *o++ = 'B';
-        }
-        *sp = newstate;
-    }
+    o = iso2022jp_put_state(sp, o, sp[0], newstate);
 
     if (l == 1) {
         *o++ = s[0] & 0x7f;
@@ -444,15 +435,26 @@ rb_cp50221_encoder = {
     iso2022jp_encoder_reset_sequence_size, finish_iso2022jp_encoder
 };
 
-RBIMPL_ATTR_NONSTRING() static const char *tbl0208 =
-    "\x21\x23\x21\x56\x21\x57\x21\x22\x21\x26\x25\x72\x25\x21\x25\x23" \
-    "\x25\x25\x25\x27\x25\x29\x25\x63\x25\x65\x25\x67\x25\x43\x21\x3C" \
-    "\x25\x22\x25\x24\x25\x26\x25\x28\x25\x2A\x25\x2B\x25\x2D\x25\x2F" \
-    "\x25\x31\x25\x33\x25\x35\x25\x37\x25\x39\x25\x3B\x25\x3D\x25\x3F" \
-    "\x25\x41\x25\x44\x25\x46\x25\x48\x25\x4A\x25\x4B\x25\x4C\x25\x4D" \
-    "\x25\x4E\x25\x4F\x25\x52\x25\x55\x25\x58\x25\x5B\x25\x5E\x25\x5F" \
-    "\x25\x60\x25\x61\x25\x62\x25\x64\x25\x66\x25\x68\x25\x69\x25\x6A" \
-    "\x25\x6B\x25\x6C\x25\x6D\x25\x6F\x25\x73\x21\x2B\x21\x2C";
+/* JIS0201 to JIS0208 conversion table */
+enum {tbl0208_num = 0xDF - 0xA1 + 1};
+RBIMPL_ATTR_NONSTRING_ARRAY() static const char tbl0208[tbl0208_num][2] = {
+    "\x21\x23", "\x21\x56", "\x21\x57", "\x21\x22",
+    "\x21\x26", "\x25\x72", "\x25\x21", "\x25\x23",
+    "\x25\x25", "\x25\x27", "\x25\x29", "\x25\x63",
+    "\x25\x65", "\x25\x67", "\x25\x43", "\x21\x3C",
+    "\x25\x22", "\x25\x24", "\x25\x26", "\x25\x28",
+    "\x25\x2A", "\x25\x2B", "\x25\x2D", "\x25\x2F",
+    "\x25\x31", "\x25\x33", "\x25\x35", "\x25\x37",
+    "\x25\x39", "\x25\x3B", "\x25\x3D", "\x25\x3F",
+    "\x25\x41", "\x25\x44", "\x25\x46", "\x25\x48",
+    "\x25\x4A", "\x25\x4B", "\x25\x4C", "\x25\x4D",
+    "\x25\x4E", "\x25\x4F", "\x25\x52", "\x25\x55",
+    "\x25\x58", "\x25\x5B", "\x25\x5E", "\x25\x5F",
+    "\x25\x60", "\x25\x61", "\x25\x62", "\x25\x64",
+    "\x25\x66", "\x25\x68", "\x25\x69", "\x25\x6A",
+    "\x25\x6B", "\x25\x6C", "\x25\x6D", "\x25\x6F",
+    "\x25\x73", "\x21\x2B", "\x21\x2C"
+};
 
 static ssize_t
 fun_so_cp50220_encoder(void *statep, const unsigned char *s, size_t l,
@@ -461,22 +463,21 @@ fun_so_cp50220_encoder(void *statep, const unsigned char *s, size_t l,
     unsigned char *output0 = o;
     unsigned char *sp = statep;
 
-    if (sp[0] == G0_JISX0201_KATAKANA) {
+    if (sp[0] == G0_JISX0201_KATAKANA && sp[2]) {
         int c = sp[2] & 0x7F;
-        const char *p = tbl0208 + (c - 0x21) * 2;
-        if (sp[1] != G0_JISX0208_1983) {
-            *o++ = 0x1b;
-            *o++ = '$';
-            *o++ = 'B';
-        }
+        const char *p = tbl0208[c - 0x21];
+        sp[2] = 0;
+        o = iso2022jp_put_state(sp, o, sp[1], G0_JISX0208_1983);
         sp[0] = G0_JISX0208_1983;
         *o++ = *p++;
         if (l == 2 && s[0] == 0x8E) {
             if (s[1] == 0xDE) {
+                /* VOICED SOUND MARK */
                 *o++ = *p + 1;
                 return o - output0;
             }
             else if (s[1] == 0xDF && (0x4A <= c && c <= 0x4E)) {
+                /* SEMI-VOICED SOUND MARK */
                 *o++ = *p + 2;
                 return o - output0;
             }
@@ -485,21 +486,25 @@ fun_so_cp50220_encoder(void *statep, const unsigned char *s, size_t l,
     }
 
     if (l == 2 && s[0] == 0x8E) {
-        const char *p = tbl0208 + (s[1] - 0xA1) * 2;
         if ((0xA1 <= s[1] && s[1] <= 0xB5) ||
             (0xC5 <= s[1] && s[1] <= 0xC9) ||
             (0xCF <= s[1] && s[1] <= 0xDF)) {
-            if (*sp != G0_JISX0208_1983) {
-                *o++ = 0x1b;
-                *o++ = '$';
-                *o++ = 'B';
-                *sp = G0_JISX0208_1983;
-            }
+            /* May not be followed by a sound mark */
+            const char *p = tbl0208[s[1] - 0xA1];
+            o = iso2022jp_put_state(sp, o, *sp, G0_JISX0208_1983);
             *o++ = *p++;
             *o++ = *p;
             return o - output0;
         }
 
+        if (s[1] > 0xDF) {      /* undef */
+            o = iso2022jp_put_state(sp, o, *sp, G0_JISX0201_KATAKANA);
+            *o++ = s[1] & 0x7F;
+            sp[2] = 0;
+            return o - output0;
+        }
+
+        /* Katakana that may be followed by a sound mark */
         sp[2] = s[1];
         sp[1] = sp[0];
         sp[0] = G0_JISX0201_KATAKANA;
@@ -519,23 +524,16 @@ finish_cp50220_encoder(void *statep, unsigned char *o, size_t osize)
     if (*sp == G0_ASCII)
         return 0;
 
-    if (sp[0] == G0_JISX0201_KATAKANA) {
+    if (sp[0] == G0_JISX0201_KATAKANA && sp[2]) {
         int c = sp[2] & 0x7F;
-        const char *p = tbl0208 + (c - 0x21) * 2;
-        if (sp[1] != G0_JISX0208_1983) {
-            *o++ = 0x1b;
-            *o++ = '$';
-            *o++ = 'B';
-        }
+        const char *p = tbl0208[c - 0x21];
+        o = iso2022jp_put_state(sp, o, sp[1], G0_JISX0208_1983);
         sp[0] = G0_JISX0208_1983;
         *o++ = *p++;
         *o++ = *p;
     }
 
-    *o++ = 0x1b;
-    *o++ = '(';
-    *o++ = 'B';
-    *sp = G0_ASCII;
+    o = iso2022jp_put_state(sp, o, sp[0], G0_ASCII);
 
     return o - output0;
 }
@@ -565,4 +563,3 @@ TRANS_INIT(iso2022)
     rb_register_transcoder(&rb_cp50220_encoder);
     rb_register_transcoder(&rb_cp50221_encoder);
 }
-
diff --git a/test/ruby/test_transcode.rb b/test/ruby/test_transcode.rb
index 4b63263aae..2509594ba6 100644
--- a/test/ruby/test_transcode.rb
+++ b/test/ruby/test_transcode.rb
@@ -1615,6 +1615,8 @@ def test_to_cp50221
     assert_equal("\e$B%*!+%,%I%J!+%N!+%P%\\%^!+%Q%]%\"\e(B".force_encoding("cp50220"),
         "\xB5\xDE\xB6\xDE\xC4\xDE\xC5\xDE\xC9\xDE\xCA\xDE\xCE\xDE\xCF\xDE\xCA\xDF\xCE\xDF\xB1".
                  encode("cp50220", "sjis"))
+    assert_equal("\e$B\x21\x23\e(I\x7E\e(B".force_encoding("cp50220"),
+                 "\x8E\xA1\x8E\xFE".encode("cp50220", "cp51932"))
   end
 
   def test_iso_2022_jp_1
